### ISSUE TITLE
[BUG] Inferring schema for a large `Dataset` is slow due to inefficient import check

### ISSUE THREAD

**Issue by justalfred** [mlflow/mlflow#13085]
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

- Client: 2.16.0
- Tracking server: N/A

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: macOS
- **Python version**: 3.12.5
- **yarn version, if running the dev UI**: N/A

### Describe the problem

Trying MLflow out for a spin, I found logging inputs to be unusably slow for large data sets. I tracked the slowdown to inferring the schema for a `Dataset`. I was working with Pandas DataFrames, but looking at the code, I believe this occurs in many other cases, too. (Spark dataframes and numpy ndarrays don't seem to be among those cases, which may be why there hasn't been much noise around the issue.) The issue comes from calls to `mlflow.types.utils._infer_colspec_type`. Specifically, [this line in the `.get_all_types` method in the `DataTypes` enum](https://github.com/mlflow/mlflow/blob/01f24530015fe97e738a819949b44211a2193f84/mlflow/types/schema.py#L123):

```python
if importlib.util.find_spec("pyspark") is not None:
```

This call costs about 25us on my machine and is run up to 8 times per data element, depending on the elements' types as ordered in `._infer_scalar_datatype`. That adds up to about 0.5 sec to infer the type of a column/list of 10,000 integers (`long`), 1.5 seconds for strings. For a reasonably sized DataFrame, this can quickly grow into hours. This may be what the user who filed #13020 was experiencing. Users can turn off `log_datasets` in `.autolog`, but the otherwise-handy `.evaluate` also logs the evaluation data set with no option to turn that off.

To confirm this call to `.find_spec` as the issue, if I change the line above to `if None is not None:`, then the time to infer a Series of 10,000 strings (calling `._infer_pandas_column`) shrinks from 1.55s to 144us.

I don't use pyspark, so I haven't tested a general fix. I assume the intention with the above is to add pyspark types iff the user is using pyspark, as indicated by whether it's been imported, not whether it *can* be imported. So you can't check this once when `mlflow` is imported since imports may happen in different orders. This makes sense. I think checking this once per call to `_infer_schema` would be ideal, but I'm unsure of a clean way to do that. One would like to set a flag on the `DataType` class, but it's an `Enum`. Python >=3.11 offers the `@nonmember` decorator that would allow this, but if I'm reading right, MLflow needs to work for python 3.8.

But I go thinking too hard about this, maybe someone else sees an obvious solution.

### Tracking information

```shell
N/A
```

### Code to reproduce issue

```
from time import time
import numpy as np
import mlflow

NUM_ELEMENTS = 10000

def time_infer_schema(data, label):
    start_time = time()
    dtype = mlflow.types.utils._infer_schema(data)
    print(label, time() - start_time)
    print(dtype)

time_infer_schema(np.arange(NUM_ELEMENTS), "ndarray")
# [Tensor('int64', (-1,))], bypasses _infer_colspec_type, fast

time_infer_schema(list(range(NUM_ELEMENTS)), "int list")
# [long (required)], 2 slow calls per element, ~.5s

time_infer_schema([str(i) for i in range(NUM_ELEMENTS)], "str list")
# [string (required)], 6 slow calls per element, ~1.5s

time_infer_schema(pd.Series(range(NUM_ELEMENTS)), "int Series")
# [long (required)], not pandas "O" type, treated as np dtype, fast

time_infer_schema(pd.Series([str(i) for i in range(NUM_ELEMENTS)]), "str Series")
# [string (required)], ~1.5s
```

### Stack trace

```
N/A
```

### Other info / logs

```
N/A
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [X] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by daniellok-db** [comment]
cc @serena-ruan or @WeichenXu123 i think this issue makes sense, the end goal is to call `importlib.util.find_spec("pyspark")` only once per schema inference, since this function is slow.

I guess we can pass in `types` as a parameter so it can be initialized during the initial `infer_schema()` call, but I don't know how messy that would make the API / how big of a refactor that would be. Maybe also possible to read from some singleton that we create during the call and restore to default after, but i think that seems a bit hacky.

**Comment by serena-ruan** [comment]
Thanks for reporting this @justalfred 
Do you mind having a test to replace the `importlib.util.find_spec` with 
```
try: 
    import pyspark 
except ImportError:
    pass
```
and see if it has a similar speed?

**Comment by serena-ruan** [comment]
What about this:
```
diff --git a/mlflow/types/schema.py b/mlflow/types/schema.py
index 794a87589..52118e7fd 100644
--- a/mlflow/types/schema.py
+++ b/mlflow/types/schema.py
@@ -18,6 +18,7 @@ ARRAY_TYPE = "array"
 OBJECT_TYPE = "object"
 MAP_TYPE = "map"
 SPARKML_VECTOR_TYPE = "sparkml_vector"
+_PYSPARK_AVAILABLE = importlib.util.find_spec("pyspark") is not None
 
 
 class DataType(Enum):
@@ -120,7 +121,7 @@ class DataType(Enum):
 
     def get_all_types(self):
         types = [self.to_numpy(), self.to_pandas(), self.to_python()]
-        if importlib.util.find_spec("pyspark") is not None:
+        if _PYSPARK_AVAILABLE:
             types.append(self.to_spark())
         if self.name == "datetime":
             types.extend([np.datetime64, dt.datetime])
```
I don't think people tend to change their environment a lot if they work with spark, pyspark is always installed. Anyways if they change the env, they can always re-import mlflow.

**Comment by justalfred** [comment]
Yep, that brings the slow lines from 1.5s to <30ms. I had misunderstood what `.find_spec` does, so my concern about determining whether `pyspark` is already imported is invalid. This seems like a good solution, and I can't think of a realistic case where it would change any behavior (aside from the speedup). @serena-ruan since it's your solution, would you like to open a PR, or would you prefer I do it and save you that work?
